File: vsh/oled_connector.py
from typing import Callable, Optional

import sim
from oled import OLED
from oled.sh1107 import Base, Cmd, DataBytes
from .i2c_connector import I2CConnector
from .shared import HIGH_STATES, MED_STATES, track

__all__ = ["OLEDConnector"]


class OLEDConnector:
    oled: OLED
    process_cb: Callable[[list[Base | DataBytes]], None]
    addr: int

    i2c_connector: I2CConnector

    def __init__(
        self,
        oled: OLED,
        process_cb: Callable[[list[Base | DataBytes]], None],
        *,
        addr: int = 0x3C,
    ):
        self.oled = oled
        self.process_cb = process_cb
        self.addr = addr

        self.i2c_connector = I2CConnector(oled.i2c, addr)

    def sim_process(self) -> sim.Generator:
        parser: Optional[Cmd.Parser] = None

        while True:
            track(HIGH_STATES, "command", (yield self.oled.i_cmd), OLED.Command)

            track(HIGH_STATES, "result", (yield self.oled.o_result), OLED.Result)
            track(MED_STATES, "remain", (yield self.oled.remain))
            track(MED_STATES, "offset", (yield self.oled.offset))

            track(MED_STATES, "rom_rd.addr", (yield self.oled.rom_rd.addr))
            track(MED_STATES, "rom_rd.data", (yield self.oled.rom_rd.data))

            byte = yield from self.i2c_connector.sim_tick()
            match byte:
                case None:
                    pass

                case "addressed":
                    parser = Cmd.Parser()

                case "error":
                    parser = None

                case "fish":
                    if not parser or not parser.valid_finish:
                        print("command parser fish without valid_finish")
                    parser = None

                case byte:
                    assert parser
                    cmds = parser.feed([byte])
                    if parser.unrecoverable:
                        print(
                            "command parser noped out, resetting with: x",
                            "".join([f"{b:02x}" for b in parser.bytes]),
                            " -- partial_cmd: x",
                            "".join([f"{b:02x}" for b in parser.partial_cmd]),
                            " -- state: ",
                            parser.state,
                            " -- continuation: ",
                            parser.continuation,
                        )
                        parser = None
                        self.i2c_connector.reset()
                    self.process_cb(cmds)

            yield

File: vsh/src/OLEDConnector.zig
const std = @import("std");

const Cxxrtl = @import("./Cxxrtl.zig");
const I2CConnector = @import("./I2CConnector.zig");
const I2CBBConnector = @import("./I2CBBConnector.zig");
const FPGAThread = @import("./FPGAThread.zig");
const Cmd = @import("./Cmd.zig");

const OLEDConnector = @This();

const InnerI2CConnector = union(enum) {
    I2CConnector: I2CConnector,
    I2CBBConnector: I2CBBConnector,
};

i2c_connector: InnerI2CConnector,

state: union(enum) {
    Unaddressed,
    AddressedWrite: Cmd.Parser,
    AddressedRead,
} = .Unaddressed,

pub const Tick = union(enum) {
    Pass,
    AddressedWrite,
    AddressedRead: *u8,
    Error,
    Fish,
    Byte: u8,
};

pub const RW = enum(u1) {
    W = 0,
    R = 1,
};

pub fn tick(self: *OLEDConnector, fpga_thread: *FPGAThread) void {
    self.tick_i2c(fpga_thread);
}

fn tick_i2c(self: *OLEDConnector, fpga_thread: *FPGAThread) void {
    switch (self.i2c_connector) {
        inline else => |*i2c_connector| {
            switch (i2c_connector.tick()) {
                .Pass => {},
                .AddressedWrite => {
                    self.state = .{ .AddressedWrite = .{} };
                },
                .AddressedRead => |byte_out| {
                    self.state = .AddressedRead;
                    const sh1107 = fpga_thread.acquire_sh1107();

                    // not busy, display on/off, ID=7
                    byte_out.* = 0x07 | (if (sh1107.power) @as(u8, 0x00) else @as(u8, 0x40));
                },
                .Error => {
                    std.debug.print("i2c error\n", .{});
                    self.state = .Unaddressed;
                },
                .Fish => {
                    switch (self.state) {
                        .Unaddressed => std.debug.print("i2c fish while unaddressed\n", .{}),
                        .AddressedWrite => |parser| if (!parser.valid_finish) {
                            std.debug.print("i2c fish without valid_finish\n", .{});
                        },
                        .AddressedRead => {},
                    }
                    self.state = .Unaddressed;
                },
                .Byte => |byte| {
                    switch (self.state) {
                        .AddressedWrite => |*parser| switch (parser.feed(byte)) {
                            .Pass => {},
                            .Unrecoverable => {
                                std.debug.print("command parser noped out, fed {x:0>2} -- " ++
                                    "state: {} / continuation: {} / partial_cmd: {?x:0>2}\n", .{
                                    byte,
                                    parser.state,
                                    parser.continuation,
                                    parser.partial_cmd,
                                });
                                self.state = .Unaddressed;
                                i2c_connector.reset();
                            },
                            .Command => |cmd| {
                                fpga_thread.process_cmd(cmd);
                            },
                            .Data => |data| {
                                fpga_thread.process_data(data);
                            },
                        },
                        else => std.debug.print("i2c got Byte while not AddressedWrite\n", .{}),
                    }
                },
            }
        },
    }
}

pub fn init(cxxrtl: Cxxrtl, addr: u7) OLEDConnector {
    var i2c_connector: InnerI2CConnector = undefined;

    if (cxxrtl.find(bool, "scl__o") != null) {
        i2c_connector = .{ .I2CConnector = I2CConnector.init(cxxrtl, addr) };
    } else {
        i2c_connector = .{ .I2CBBConnector = I2CBBConnector.init(cxxrtl, addr) };
    }

    return .{
        .i2c_connector = i2c_connector,
    };
}
